### ISSUE TITLE
Design and Implement Random Skills

### ISSUE THREAD

**Issue by anchristensen** [uchicago-cs/chiventure#1460]
Design and implement random skills, another complex skill type which has an element of randomness to determine how, or if, it is executed. For instance, one skill could be "All-Out Attack: Deal 100 damage; misses 30% of the time." This effect succeeds or fails randomly. Another example would be "Chaos Beam: Equal chance to Stun, Freeze, or Burn target." This executes one of three potential outcomes. This will involve adding functions and documentation supporting these skills, and possibly adding some effects, such as random chance and/or random range effect. We will also need to evaluate how best to support a skill such as the chaos beam, which may require nesting or another sort of supporting effect.

There is further description of random skills here: https://github.com/uchicago-cs/chiventure/wiki/Skill-Trees-~-Integrating-complex-skill-(combined,-random,-sequential,-etc.)-implementation

team members: Annika, Arya, Chelsea

**Comment by anchristensen** [comment]
Brief update: the designing is complete. We have outlined what precisely needs to be done to implement this issue. Specifically, we will need to: 1) add RANDOM as an enum option to complex_skill_type_t once the complex_skill pull request has been approved. 2) add a random effect for random chance. This will work for skills like "All-Out Attack" as described in the opening statement; it becomes Chance(70%) in sequence with Damage(100) with this effect. 3) add a random effect for a random range. When it executes, it generates a number between the range bounds, and passes that result forward so that, for instance, "Flex: Gain 10 strength-up for 3-5 turns." is supported. 4) implement a switch block for effects like "Equal chance to Stun, Freeze, or Burn target." This will have an equal chance to choose one of those options. 5) an execute function which will take a new enum type to tell which of the three random types it is, and then executes it.

**Comment by anchristensen** [comment]
Update: we decided to add three specific enums to the complex_skill_type_t struct (RANDOM_CHANCE, RANDOM_RANGE, RANDOM_SWITCH) instead of just one RANDOM enum so that we can work within the complex_skill_t struct for all of the effects. We also decided on three new structs, one for each of the random effect types, so that that information can easily be passed into the execution functions. The execution of chance and range effect functions have been implemented. We discussed how to implement the switch block and decided on a new struct type, a random_switch_type_t which will take in the complex skill and also an array of the chances for each subskill to be executed. This will allow the game developers to pass in the specific percentages they want for each subskill to be executed (ex: stun is 10%, burn 60%, freeze 30% instead of always being equal chances). Furthermore we came up with a question to ask about in the sprint meeting tomorrow; we want to ask about how to test these skills since they are fundamentally random so outcomes may be different than expected. Finally we discussed how to add these execute functions within the general complex_skill_execute function since they have varying parameters--we are attempting to use variable arguments as described here https://www.tutorialspoint.com/cprogramming/c_variable_arguments.htm but may have to reexamine this if it doesn't work and/or leave it for the future to implement a more generalized execute function for random skills

**Comment by anchristensen** [comment]
Update: All three execute functions have been written and are compiling. For now, we have decided against a generalized execute function which means the user will have to use the specific execute_random_range or execute_random_switch (etc.). This general function will be left for future implementation--a union might be able to be used, for example, but for now we don't see this as crucial to this issue and our main focus is getting these functions up and running. Therefore the only things left to do are to add testing and write new init and free functions for the new random structs, and we will hopefully get the PR in soon.

**Comment by anchristensen** [comment]
Closing statement: With pull request #1641, this issue will be finished. In this issue, we designed and implemented a new type of complex skills, random skills as initially described here: https://github.com/uchicago-cs/chiventure/wiki/Skill-Trees-~-Integrating-complex-skill-(combined,-random,-sequential,-etc.)-implementation. This involved creating three new structs, one for each type of random skill (chance, range, switch) which contain a complex skill as well as individualized parameters such as, for random chance, a float with the chance of failure. Then, we wrote new, init, and free functions for each of these structs, as well as execute functions for each (which also involved writing some random generation code using the C rand() function, which may not be as perfectly random as a third party library, but works for now). Finally we added testing for all of these functions. I wrote the three execute functions, the random_switch and random_range new, init, and free functions, and testing for the execute functions. Arya wrote testing for all of the new init and free functions. Chelsea wrote the new, init, and free functions for random_chance.
